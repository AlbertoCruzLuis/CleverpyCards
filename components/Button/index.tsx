import classNames from "classnames"
import type { HTMLProps, MouseEventHandler } from "react"
import { forwardRef } from "react"

export type ButtonVariant =
  | "primary"
  | "secondary"
  | "tertiary"

export type ButtonSize = "large" | "small"

export type ButtonProps = Omit<HTMLProps<HTMLButtonElement>, "size"> & {
  children?: React.ReactNode
  className?: string
  onClick?: MouseEventHandler<HTMLButtonElement>
  variant?: ButtonVariant
  size?: ButtonSize
  icon?: any
}

const buttonVariant = {
  primary: "bg-pink-1000 hover:bg-pink-500 text-white",
  secondary: "border border-pink-1000 hover:border-pink-400 hover:text-pink-400 text-pink-1000",
  tertiary: "text-white"
}

const buttonSize = {
  large: "px-8 py-2",
  small: "px-4 py-2"
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button (
    {
      children,
      className,
      onClick,
      variant = "primary",
      size = "small",
      icon
    }: ButtonProps,
    ref
  ) {
    const cn = classNames(className, {
      [buttonVariant.primary]: variant === "primary",
      [buttonVariant.secondary]: variant === "secondary",
      [buttonVariant.tertiary]: variant === "tertiary",
      [buttonSize.large]: size === "large",
      [buttonSize.small]: size === "small"
    })

    return (
      <button
        className={cn}
        onClick={onClick}
        type="button"
        ref={ref}
      >
        {icon && icon}
        {children}
      </button>
    )
  }
)
