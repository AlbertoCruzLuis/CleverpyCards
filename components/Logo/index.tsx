import CleverpyIcon from "assets/icons/DrawArrow"
import { FC, memo } from "react"

export interface ILogoProps {
  width?: string,
  height?: string
}

const LogoComponent: FC<ILogoProps> = ({
  width = "50",
  height = "50"
}) => {
  return (
    <div className="flex items-center gap-1">
      <CleverpyIcon
        width={width}
        height={height}
      />
    </div>
  )
}

export const Logo = memo(LogoComponent)
