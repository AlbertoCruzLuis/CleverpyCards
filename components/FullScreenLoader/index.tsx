import { Background } from "components/Background"
import { useAuth } from "hooks/useAuth"
import { useRouter } from "next/router"
import { useEffect, useState } from "react"
import ScaleLoader from "react-spinners/ScaleLoader"

interface IFullScreenLoaderProps {
  children: React.ReactNode
}

export const FullScreenLoader = ({ children }: IFullScreenLoaderProps) => {
  const { isAuth } = useAuth()
  const [isLoading, setLoading] = useState(true)
  const router = useRouter()

  const showLoader = () => {
    const isLoginPage = router.pathname === "/login"
    const isShowLoader = isAuth === isLoginPage
    setLoading(isShowLoader)
  }

  useEffect(() => {
    setTimeout(() => {
      showLoader()
    }, 800)
  }, [router, isAuth])

  if (!isLoading) {
    return (
      <>
        {children}
      </>
    )
  }

  return (
    <Background>
      <div className="flex items-center justify-center h-screen">
        <ScaleLoader color="#fff" />
      </div>
    </Background>
  )
}
