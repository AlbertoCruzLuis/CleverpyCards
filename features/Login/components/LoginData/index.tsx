import { Button } from "components/Button"
import { ErrorMessage, Field, useFormikContext } from "formik"
import toast from "react-hot-toast"

export const styles = {
  firstLabel: "text-gray-200 text-sm font-semibold mb-2 block",
  label: "text-neutral-200 text-sm font-semibold mb-2 block",
  field: "bg-neutral-100 text-gray-700 focus:outline-none focus:shadow-outline border border-gray-300 rounded py-2 px-4 block w-full appearance-none",
  socialField: "bg-neutral-100 text-gray-700 focus:outline-none focus:shadow-outline border border-gray-300 rounded-r-md py-2 px-4 block w-full appearance-none",
  errorMessage: "text-red-500 text-sm"
}

export const LoginData = () => {
  const { handleSubmit, errors } = useFormikContext()

  const submitForm = async () => {
    try {
      const isError = Object.keys(errors).length !== 0

      if (isError) {
        toast.error("Please fill all required fields")
      }

      handleSubmit()
    } catch (error) {
      console.error(error)
    }
  }

  return (
    <div className="flex flex-col items-center gap-10 bg-[#1D0327] p-8 rounded-md">
      <div className="flex flex-col gap-4">
        <div>
          <label className={styles.label} htmlFor="email">Email</label>
          <Field id="email" className={styles.field} type="text" name="email" placeholder="alberto@cleverpy.com" />
          <ErrorMessage className={styles.errorMessage} name="email" component="div" />
        </div>
        <div>
          <label className={styles.label} htmlFor="password">Password</label>
          <Field id="password" className={styles.field} type="password" name="password" placeholder="alberto" />
          <ErrorMessage className={styles.errorMessage} name="password" component="div" />
        </div>
      </div>
      <Button
        className="w-full rounded-sm"
        variant="primary"
        onClick={submitForm}
      >
        <span>Sign in</span>
      </Button>
    </div>
  )
}
