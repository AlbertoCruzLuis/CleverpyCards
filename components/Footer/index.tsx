import { CustomLink } from "components/CustomLink"
import { Logo } from "components/Logo"
import { BsGithub, BsTwitter } from "react-icons/bs"

export const Footer = () => {
  const year = new Date().getFullYear()
  return (
    <footer className="p-8 bg-[#1D0327]">
      <div className="mx-auto xl:container">
        <div className="grid grid-cols-4 gap-8 xs:grid-cols-1 sm:grid-cols-2 md:grid-cols-2">
          <div className="flex flex-col gap-4">
            <CustomLink href="/">
              <Logo width="90" height="90" />
            </CustomLink>
            <p className="max-w-sm font-semibold text-white">Cleverpy Cards</p>
            <div className="flex gap-4">
              <CustomLink href="" className="p-2 border-2 border-neutral-700">
                <BsTwitter size="1.2rem" className="transition-all text-neutral-400 hover:text-white" />
              </CustomLink>
              <CustomLink href="" className="p-2 border-2 border-neutral-700">
                <BsGithub size="1.2rem" className="transition-all text-neutral-400 hover:text-white" />
              </CustomLink>
            </div>
          </div>
          <div>
            <h2 className="mb-6 text-sm font-semibold text-white uppercase">Resources</h2>
            <ul className="text-neutral-400">
              <li className="mb-4">
                <CustomLink href="/" className="hover:underline">Home</CustomLink>
              </li>
            </ul>
          </div>
          <div>
            <h2 className="mb-6 text-sm font-semibold text-white uppercase">Follow us</h2>
            <ul className="text-neutral-400">
              <li className="mb-4">
                <CustomLink href="" className="hover:underline ">Github</CustomLink>
              </li>
              <li>
                <CustomLink href="" className="hover:underline">Twitter</CustomLink>
              </li>
            </ul>
          </div>
          <div>
            <h2 className="mb-6 text-sm font-semibold text-white uppercase">Legal</h2>
            <ul className="text-neutral-400">
              <li className="mb-4">
                <CustomLink href="/" className="hover:underline">Privacy Policy</CustomLink>
              </li>
              <li>
                <CustomLink href="#" className="hover:underline">Terms &amp; Conditions</CustomLink>
              </li>
            </ul>
          </div>
        </div>
        <hr className="my-6 border-neutral-800/50 sm:mx-auto lg:my-8" />
        <div className="sm:flex sm:items-center sm:justify-between">
          <span className="text-sm text-neutral-400 sm:text-center">© {year} <CustomLink href="/" className="hover:underline">Cleverpy™</CustomLink>. All Rights Reserved.
          </span>
        </div>
      </div>
    </footer>
  )
}
