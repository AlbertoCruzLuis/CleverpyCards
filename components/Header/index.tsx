import { useScrollPosition } from "@n8tb1t/use-scroll-position"
import { CustomLink } from "components/CustomLink"
import { ILogoProps } from "components/Logo"
import { useUI } from "hooks/useUI"
import dynamic from "next/dynamic"
import { useRouter } from "next/router"
import { useEffect, useState } from "react"
import { BiUser } from "react-icons/bi"
import { useWindowSize } from "react-use"

const Logo = dynamic<ILogoProps>(
  () => import("components/Logo").then(mod => mod.Logo),
  { ssr: false }
)

export const Header = () => {
  const [isOpen, setOpen] = useState(false)
  const { width } = useWindowSize()
  const [isUpScroll, setUpScroll] = useState(true)
  const [positionScroll, setPositionScroll] = useState(0)
  const router = useRouter()
  const { openSidebar } = useUI()

  useEffect(() => {
    setTimeout(() => {
      setOpen(false)
    }, 1000)
  }, [router.asPath, location.href])

  useEffect(() => {
    document.body.style.overflowY = isOpen ? "hidden" : "scroll"
  }, [isOpen])

  useEffect(() => {
    width > 640 && setOpen(false)
  }, [width])

  useScrollPosition(
    ({ prevPos, currPos }) => {
      const isShow = currPos.y > prevPos.y
      if (isShow !== isUpScroll) setUpScroll(isShow)
      setPositionScroll(currPos.y)
    },
    [positionScroll]
  )

  return (
    <header className={`${positionScroll > -100 && "bg-transparent"} transition-all bg-[#1D0327] sticky top-0 z-20 px-8 py-4`}>
      <div className="flex items-center justify-between mx-auto xl:container">
        <div className="flex divide-x divide-gray-500">
          <CustomLink className="pr-5" href="/">
            <Logo />
          </CustomLink>
        </div>
        <div className="">
          <div className="flex gap-8">
            <button className="relative" onClick={openSidebar}>
              <BiUser size="1.2rem" className="text-neutral-200" />
            </button>
          </div>
        </div>
      </div>
    </header>
  )
}
