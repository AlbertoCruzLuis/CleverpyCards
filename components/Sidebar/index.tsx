import { Button } from "components/Button"
import { Logo } from "components/Logo"
import { useAuth } from "hooks/useAuth"
import { useUI } from "hooks/useUI"
import { SidebarLayout } from "layouts/SidebarLayout"

export const Sidebar = () => {
  const { logOut } = useAuth()
  const { closeSidebar } = useUI()

  const handleLogOut = () => {
    logOut()
    closeSidebar()
  }

  return (
    <SidebarLayout>
      <div className="flex flex-col items-center justify-center flex-1">
        <Logo width="120" height="120" />
        <Button variant="primary" onClick={handleLogOut}>
          <span>Log out</span>
        </Button>
      </div>
    </SidebarLayout>
  )
}
