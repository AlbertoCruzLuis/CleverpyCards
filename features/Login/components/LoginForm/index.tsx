import { Form, Formik } from "formik"
import { useAuth } from "hooks/useAuth"
import { loginSchema } from "utils/validation/loginSchema"
import { LoginData } from "../LoginData"

export interface IFormValues {
  email: string,
  password: string
}

export const LoginForm = () => {
  const { logIn } = useAuth()
  const initialFormValues = {
    email: "",
    password: ""
  }

  const handleSubmit = async (values: IFormValues) => {
    const formatedData = {
      email: values.email,
      password: values.password
    }

    logIn(formatedData)
  }
  return (
    <div>
      <Formik
        initialValues={initialFormValues}
        validationSchema={loginSchema}
        onSubmit={(values) => handleSubmit(values)}
      >
        {() => {
          return (
            <Form>
              <LoginData />
            </Form>
          )
        }}
      </Formik>
    </div>
  )
}
