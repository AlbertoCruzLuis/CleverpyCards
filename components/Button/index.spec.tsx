import { Button } from "components/Button"
import { AiOutlinePlusCircle } from "react-icons/ai"
import { customRender } from "test"

describe("Button Primary", () => {
  it("renders correctly", () => {
    const { container } = customRender(<Button variant="primary">Primary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders correctly with size large", () => {
    const { container } = customRender(<Button variant="primary" size="large">Primary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders correctly with icon", () => {
    const { container } = customRender(<Button variant="primary" icon={<AiOutlinePlusCircle />}>Primary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })
})

describe("Button Secondary", () => {
  it("renders correctly", () => {
    const { container } = customRender(<Button variant="secondary">Secondary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders correctly with size large", () => {
    const { container } = customRender(<Button variant="secondary" size="large">Secondary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders correctly with icon", () => {
    const { container } = customRender(<Button variant="secondary" icon={<AiOutlinePlusCircle />}>Secondary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })
})

describe("Button Tertiary", () => {
  it("renders correctly", () => {
    const { container } = customRender(<Button variant="tertiary">Tertiary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders correctly with size large", () => {
    const { container } = customRender(<Button variant="tertiary" size="large">Tertiary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders correctly with icon", () => {
    const { container } = customRender(<Button variant="tertiary" icon={<AiOutlinePlusCircle />}>Tertiary Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })
})
