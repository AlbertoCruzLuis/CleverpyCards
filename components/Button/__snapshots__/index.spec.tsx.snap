// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Primary renders correctly 1`] = `
<button
  class="bg-pink-1000 hover:bg-pink-500 text-white px-4 py-2"
>
  Primary Button
</button>
`;

exports[`Button Primary renders correctly with icon 1`] = `
<button
  class="bg-pink-1000 hover:bg-pink-500 text-white px-4 py-2"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 1024 1024"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
    />
    <path
      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
    />
  </svg>
  Primary Button
</button>
`;

exports[`Button Primary renders correctly with size large 1`] = `
<button
  class="bg-pink-1000 hover:bg-pink-500 text-white px-8 py-2"
>
  Primary Button
</button>
`;

exports[`Button Secondary renders correctly 1`] = `
<button
  class="border border-pink-1000 hover:border-pink-400 hover:text-pink-400 text-pink-1000 px-4 py-2"
>
  Secondary Button
</button>
`;

exports[`Button Secondary renders correctly with icon 1`] = `
<button
  class="border border-pink-1000 hover:border-pink-400 hover:text-pink-400 text-pink-1000 px-4 py-2"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 1024 1024"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
    />
    <path
      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
    />
  </svg>
  Secondary Button
</button>
`;

exports[`Button Secondary renders correctly with size large 1`] = `
<button
  class="border border-pink-1000 hover:border-pink-400 hover:text-pink-400 text-pink-1000 px-8 py-2"
>
  Secondary Button
</button>
`;

exports[`Button Tertiary renders correctly 1`] = `
<button
  class="text-white px-4 py-2"
>
  Tertiary Button
</button>
`;

exports[`Button Tertiary renders correctly with icon 1`] = `
<button
  class="text-white px-4 py-2"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 1024 1024"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
    />
    <path
      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
    />
  </svg>
  Tertiary Button
</button>
`;

exports[`Button Tertiary renders correctly with size large 1`] = `
<button
  class="text-white px-8 py-2"
>
  Tertiary Button
</button>
`;
